_posts/2016-11-09-Rust-1.13.md

```diff
@@ -0,0 +1,375 @@
+solving one simple problem. To illustrate, imagine we had some code to read some
+data from a file:
+
+```rust
+fn read_username_from_file() -> Result<String, io::Error> {
+    let f = File::open("username.txt");
+
+    let mut f = match f {
+        Ok(file) => file,
+        Err(e) => return Err(e),
+    };
+
+    let mut s = String::new();
+
+    match f.read_to_string(&mut s) {
+        Ok(_) => Ok(s),
+        Err(e) => Err(e),
+    }
+}
+```
+
+This code has two paths that can fail, opening the file and reading the data
+from it. If either of these fail to work, we'd like to return an error from
+`read_username_from_file`. Doing so involves `match`ing on the result of the I/O
+operations. In simple cases like this though, where we are only propagating
+errors up the call stack, the matching is just boilerplate - seeing it written
+out, in the same pattern every time, doesn't provide the reader with a great
+deal of useful information.
+
+With `?`, the above code looks like this:
+
+```rust
+fn read_username_from_file() -> Result<String, io::Error> {
+    let mut f = File::open("username.txt")?;
+    let mut s = String::new();
+
+    f.read_to_string(&mut s)?;
+
+    Ok(s)
+}
+```
+
+The `?` is shorthand for the entire match statements we wrote earlier. In other
+words, `?` applies to a `Result` value, and if it was an `Ok`, it unwraps it and
+gives the inner value. If it was an `Err`, it returns from the function you're
+currently in.  Visually, it is much more straightforward. Instead of an entire
+match statement, now we are just using the single "?" character to indicate that
+here we are handling errors in the standard way, by passing them up the
+call stack.
+
+Seasoned Rustaceans may recongize that this is the same as the `try!` macro
+that's been available since Rust `1.0`. And indeed, they are the same. Before
+1.13, `read_username_from_file` could have been implemented like this:
+
+```rust
+fn read_username_from_file() -> Result<String, io::Error> {
+    let mut f = try!(File::open("username.txt"));
+    let mut s = String::new();
+
+    try!(f.read_to_string(&mut s));
+
+    Ok(s)
+}
+```
+
+So why extend the language when we already have a macro? There are multiple
+reasons. First, `try!` has proved to be extremely useful, and is used often in
+idiomatic Rust. It is used so often that we think it's worth having a sweet
+syntax. This sort of evolution is one of the great advantages of a powerful
+macro system: speculative extensions to the language syntax can be prototyped
+and iterated on without modifying the langue itself, and in return, macros that
+turn out to be especially useful can indicate missing language features. This
+evolution, from `try!` to `?` is a great example.
+
+One of the reasons `try!` needs a sweeter syntax is that it is quite
+unattractive when multiple invocations of `try!` are used in
+succession. Consider:
+
+```rust
+try!(try!(try!(foo()).bar()).baz())
```

Comment:
people asked in my PR why I did it this way, it's because it was from the justification in the RFC. If nobody writes code this way, then the RFC is poorly justified 😉